File: bridge/src/main/java/com/example/bridge/geofence/feature/GeofenceManager.java
package com.example.bridge.geofence.feature;

import static com.amap.api.fence.GeoFenceClient.GEOFENCE_IN;
import static com.amap.api.fence.GeoFenceClient.GEOFENCE_OUT;
import static com.amap.api.fence.GeoFenceClient.GEOFENCE_STAYED;

import android.content.Context;

import com.amap.api.fence.GeoFenceClient;
import com.amap.api.fence.GeoFenceListener;
import com.amap.api.location.DPoint;
import com.example.common.geofence.GeofenceConstants;

public class GeofenceManager {

    public static final String GEOFENCE_BROADCAST_ACTION = GeofenceConstants.GEOFENCE_BROADCAST_ACTION;

    private GeoFenceClient mGeoFenceClient;

    public GeofenceManager(Context context) {
        mGeoFenceClient = new GeoFenceClient(context);
        // public static final int GEOFENCE_IN      进入地理围栏
        // public static final int GEOFENCE_OUT     退出地理围栏
        // public static final int GEOFENCE_STAYED  停留在地理围栏内10分钟
        mGeoFenceClient.setActivateAction(GEOFENCE_IN|GEOFENCE_OUT|GEOFENCE_STAYED);
        mGeoFenceClient.createPendingIntent(GEOFENCE_BROADCAST_ACTION);

        // 删除所有围栏 mGeoFenceClient.removeGeoFence();
    }

    public Status CreateKeywordGeofence(Context context, String keyword /* 行政区划关键字 */, String custom) {
        if (custom.equals(GeofenceStatusManager.getFenceCustomId(context))) return Status.REPEATED_CREATION;
        // 关键字合法性校验
        mGeoFenceClient.addGeoFence(keyword, custom);
        return Status.SUCCESS;
    }

    public Status CreateLatLngGeofence(Context context, DPoint latLng, float radius /* 0(3000) - 50000m */, String custom) {
        if (custom.equals(GeofenceStatusManager.getFenceCustomId(context))) return Status.REPEATED_CREATION;
        mGeoFenceClient.addGeoFence(latLng, radius, custom);
        return Status.SUCCESS;
    }

    public void setFenceListener(GeoFenceListener listener) { // 围栏创建回调
        mGeoFenceClient.setGeoFenceListener(listener);
    }

    public enum Status {
        REPEATED_CREATION, // 重复创建
        SUCCESS,
        FAILURE,
    }
}

File: cognitive/src/main/java/geofence/manager/GeofenceManager.kt
package geofence.manager

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.amap.api.fence.GeoFenceClient
import com.amap.api.fence.GeoFenceListener
import com.amap.api.location.DPoint
import com.example.common.geofence.GeofenceConstants
import com.example.common.geofence.model.BarrierInfo
import androidx.core.content.edit

class GeofenceManager(context: Context) {

    companion object {
        private const val TAG = "CognitiveGeofenceMgr"
        const val GEOFENCE_BROADCAST_ACTION = GeofenceConstants.GEOFENCE_BROADCAST_ACTION
        private const val SP_NAME = "cognitive_geofence_status"
        private const val KEY_ELDERNAME = "eldername"
        private const val KEY_LAT = "lat"
        private const val KEY_LON = "lon"
        private const val KEY_RADIUS = "radius"
    }

    private val context: Context = context.applicationContext
    private val geoFenceClient: GeoFenceClient = GeoFenceClient(context)
    private val sp: SharedPreferences = context.getSharedPreferences(SP_NAME, Context.MODE_PRIVATE)

    private var currentFenceId: String? = null

    init {
        geoFenceClient.setActivateAction(
            GeoFenceClient.GEOFENCE_IN or
            GeoFenceClient.GEOFENCE_OUT or
            GeoFenceClient.GEOFENCE_STAYED
        )
        geoFenceClient.createPendingIntent(GEOFENCE_BROADCAST_ACTION)
    }

    fun setFenceListener(listener: GeoFenceListener) {
        geoFenceClient.setGeoFenceListener(listener)
    }

    fun createGeofence(barrierInfo: BarrierInfo): Boolean {
        val customId = barrierInfo.eldername
        if (customId == currentFenceId) {
            Log.w(TAG, "Geofence already created for: $customId")
            return false
        }

        val center = DPoint(barrierInfo.lat, barrierInfo.lon)
        val radius = barrierInfo.radius.toFloat()

        geoFenceClient.addGeoFence(center, radius, customId)
        currentFenceId = customId

        saveBarrierInfo(barrierInfo)

        Log.d(TAG, "Geofence created: eldername=$customId, lat=${barrierInfo.lat}, lon=${barrierInfo.lon}, radius=$radius")
        return true
    }

    fun restoreGeofenceIfExists(): Boolean {
        val eldername = sp.getString(KEY_ELDERNAME, null) ?: return false
        val lat = sp.getFloat(KEY_LAT, 0f).toDouble()
        val lon = sp.getFloat(KEY_LON, 0f).toDouble()
        val radius = sp.getFloat(KEY_RADIUS, 0f).toDouble()

        if (lat == 0.0 && lon == 0.0) {
            return false
        }

        val barrierInfo = BarrierInfo(eldername, lon, lat, radius)
        Log.d(TAG, "Restoring geofence from local: $barrierInfo")
        return createGeofence(barrierInfo)
    }

    private fun saveBarrierInfo(barrierInfo: BarrierInfo) {
        sp.edit {
            putString(KEY_ELDERNAME, barrierInfo.eldername)
                .putFloat(KEY_LAT, barrierInfo.lat.toFloat())
                .putFloat(KEY_LON, barrierInfo.lon.toFloat())
                .putFloat(KEY_RADIUS, barrierInfo.radius.toFloat())
        }
        Log.d(TAG, "BarrierInfo saved to local")
    }

    fun getSavedBarrierInfo(): BarrierInfo? {
        val eldername = sp.getString(KEY_ELDERNAME, null) ?: return null
        val lat = sp.getFloat(KEY_LAT, 0f).toDouble()
        val lon = sp.getFloat(KEY_LON, 0f).toDouble()
        val radius = sp.getFloat(KEY_RADIUS, 0f).toDouble()

        if (lat == 0.0 && lon == 0.0) {
            return null
        }
        return BarrierInfo(eldername, lon, lat, radius)
    }

    fun removeAllGeofences() {
        geoFenceClient.removeGeoFence()
        currentFenceId = null
        clearLocalStorage()
        Log.d(TAG, "All geofences removed")
    }

    fun removeGeofence(customId: String) {
        geoFenceClient.removeGeoFence()
        if (currentFenceId == customId) {
            currentFenceId = null
        }
        Log.d(TAG, "Geofence removed: $customId")
    }

    private fun clearLocalStorage() {
        sp.edit { clear() }
        Log.d(TAG, "Local storage cleared")
    }
}
